### PR TITLE
Run mode behavior re-work (#445).

### PR DESCRIPTION
**Pull request by pq**
Essentially treats run like debug without breakpoints (as described in #445).

Far from an ideal solution but enough to get us experimenting with a new UX.

/cc @devoncarew

### CODE REVIEW

File: src/io/flutter/run/daemon/RunMode.java [hunk 1]
```diff
@@ -5,22 +5,34 @@
...
 
+import org.jetbrains.annotations.NonNls;
+
 /**
  * Associate a Java name with the strings passed to Flutter that determine execution mode.
  */
 public enum RunMode {
 
+  @NonNls
   RELEASE("release"),
```
Comment (devoncarew) [inline src/io/flutter/run/daemon/RunMode.java]: Can you delete the unreferenced enums here?

File: src/io/flutter/run/daemon/RunMode.java [hunk 2]
```diff
@@ -30,6 +42,9 @@ public String mode() {
     return myModeString;
   }
 
+  /**
+   * Returns true if this is a reload/restart enabled mode (run|debug).
+   */
   public boolean isDebug() {
```
Comment (devoncarew) [inline src/io/flutter/run/daemon/RunMode.java]: Perhaps we should rename this? `isReloadEnabled`?

File: src/io/flutter/run/FlutterDebugProcess.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    if (!isDebuggingSession()) {
+
+      // Remove all but specified actions.
+      final AnAction[] leftActions = leftToolbar.getChildActionsOrStubs();
+      // Not all on the classpath so we resort to Strings.
+      final List<String> actionClassNames = Arrays
+        .asList("com.intellij.execution.actions.StopAction", "com.intellij.ui.content.tabs.PinToolwindowTabAction",
+                "com.intellij.execution.ui.actions.CloseAction", "com.intellij.ide.actions.ContextHelpAction");
+      for (AnAction a : leftActions) {
+        System.out.println(a.getClass().getName());
```
Comment (devoncarew) [inline src/io/flutter/run/FlutterDebugProcess.java]: debugging output

File: src/io/flutter/run/FlutterDebugProcess.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.lang.reflect.InvocationTargetException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+
+public class FlutterDebugProcess extends DartVmServiceDebugProcessZ {
```
Comment (devoncarew) [inline src/io/flutter/run/FlutterDebugProcess.java]: Is this class only used for the 'run' button or for both run and debug? Can you document this, and update the README.md for this directory?

File: src/io/flutter/run/FlutterRunner.java [hunk 5]
```diff
@@ -69,9 +84,97 @@ public void sessionResumed() {
...
+
+    final XDebuggerManager debuggerManager = XDebuggerManager.getInstance(env.getProject());
+    final XDebugSession debugSession = debuggerManager.startSession(env, new XDebugProcessStarter() {
+      @Override
+      @NotNull
+      public XDebugProcess start(@NotNull final XDebugSession session) {
+        final DartUrlResolver dartUrlResolver = getDartUrlResolver(env.getProject(), contextFileOrDir);
+        return new FlutterDebugProcess(session,
+                                       StringUtil.notNullize(debuggingHost, "localhost"),
+                                       observatoryPort,
```
Comment (devoncarew) [inline src/io/flutter/run/FlutterRunner.java]: FYI, going forward, everything taking an observatory port will need to be changed to taking a service protocol URL.

File: src/io/flutter/run/daemon/RunMode.java (hunk 1 above)
Comment (pq) [inline src/io/flutter/run/daemon/RunMode.java]: Sure.

File: src/io/flutter/run/daemon/RunMode.java (hunk 2 above)
Comment (pq) [inline src/io/flutter/run/daemon/RunMode.java]: Done!

File: src/io/flutter/run/FlutterDebugProcess.java (hunk 3 above)
Comment (pq) [inline src/io/flutter/run/FlutterDebugProcess.java]: Ha.  Good catch!  Done.

**Comment by pq** [comment]
Comments addressed.  PTAL.

**Comment by devoncarew** [comment]
one more tweak; lgtm

File: src/io/flutter/run/daemon/FlutterDaemonService.java [hunk 6]
```diff
@@ -183,8 +183,8 @@ public FlutterApp startApp(@NotNull Project project,
                              @NotNull RunMode mode,
                              @Nullable String relativePath)
     throws ExecutionException {
-    boolean isPaused = mode.isDebug();
-    boolean isHot = mode.isDebug() ? HOT_MODE_DEFAULT : HOT_MODE_RELEASE;
+    boolean isPaused = mode.isReloadEnabled();
```
Comment (devoncarew) [inline src/io/flutter/run/daemon/FlutterDaemonService.java]: Sorry, this pause is for the `--start-paused` flag, when starting apps. It should only be true for debugging configurations.

Comment (pq) [inline src/io/flutter/run/daemon/FlutterDaemonService.java]: Got it.  The variable could be better named (here and in `AppStart`).  Having done that we really want:

    startPaused = mode == RunMode.DEBUG;

👍

Comment (pq) [inline src/io/flutter/run/daemon/FlutterDaemonService.java]: (Done.)

**Comment by pq** [comment]
Thanks for all the back and forth.  Landing and we can iterate.

**Pull request merged**